src/log.action.ts

```diff
@@ -108,83 +72,143 @@ export const createLogAction =
 
     const persistLog = createPersistLogAction(request, context, options);
 
-    if (request.params.action === 'bulkDelete') {
+    if (action.isBulkType()) {
       await Promise.all(
         (records || []).map(async record => {
-          const recordId = (await record.params).id;
-          await persistLog(
+          const recordId = record.id();
+          await persistLog({
             recordId,
             record,
-            context.initialRecords.find(r => {
-              return r.params.id === recordId;
-            })
-          );
+            initialRecord: initialRecords.find(r => {
+              return r.id() === recordId;
+            }),
+          });
         })
       );
+    } else if (action.isResourceType()) {
+      let recordsIds = (response.records ?? []).map(r => r.id ?? r.params.id);
+      if (!recordsIds?.length) {
+        recordsIds = (records ?? []).map(r => r.id());
+      }
+      let recordId = response.record?.id ?? response.record?.params?.id;
+      if (!recordId) {
+        recordId = record?.id?.();
+      }
+
+      if (recordsIds?.length) {
+        await Promise.all(
+          recordsIds.map(async id => {
+            return persistLog({
+              recordId: id,
+              record: (records ?? []).find(r => r.id() === id),
+              initialRecord: initialRecords.find(r => {
+                return r.id() === id;
+              }),
+            });
+          })
+        );
+      }
+
+      if (recordId) {
+        await persistLog({
+          recordId,
+          record,
+          initialRecord,
+        });
+      }
     } else {
-      const recordId =
-        params.recordId ||
-        (typeof record?.params?.id === 'string'
-          ? record?.params?.id
-          : record?.params?.id?.());
+      // record action
+      // Record action with ":recordId" param in request url
+      let { recordId } = params;
+
+      // Record action should have record in context
+      if (!recordId) {
+        recordId = record?.id?.();
```

Comment:
Why not `record.id()`?

Reply:
`record` is typed as `BaseRecord | undefined` in `ActionContext`. Additionally if it's used for custom actions, the developer may modify `context.record` in an odd way so that `record.id()` won't work. Typescript wouldn't allow this probably, but a lot of people still use plain js